src/core/ghosts.cpp

```diff
@@ -221,26 +222,41 @@ void prepare_send_buffer(GhostCommunication *gc, int data_parts) {
   }
 }
 
-static void prepare_ghost_cell(Cell *cell, int size) {
-  if (ghosts_have_bonds) {
-    // free all allocated information, will be resent
-    {
-      int np = cell->n;
-      Particle *part = cell->part;
-      for (int p = 0; p < np; p++) {
-        free_particle(part + p);
-      }
+static void prepare_ghost_cell(Cell *cell, int size)  {
+  using Utils::make_span;
+  auto const old_cap = cell->max;
+
+  /* reset excess particles */
+  if(size < cell->max) {
+    for(auto &p : make_span<Particle>(cell->part + size, cell->max - size)) {
+      p = Particle{};
     }
   }
+
+  /* Adapt size */
   cell->resize(size);
-  // invalidate pointers etc
-  {
-    int np = cell->n;
-    Particle *part = cell->part;
-    for (int p = 0; p < np; p++) {
-      auto *pt = new (&part[p]) Particle();
-      pt->l = ParticleLocal{/* .ghost */ true};
-    }
+
+  /* initialize new particles */
+  if(old_cap < cell->max) {
+    auto new_parts = make_span(cell->part + old_cap, cell->max - old_cap);
+    std::uninitialized_fill(new_parts.begin(), new_parts.end(), Particle{});
+  }
+
+  for(auto &p : Utils::make_span(cell->part, cell->n)) {
+    /* Keep heap if any, but set the logical size to 0 */
```

Comment:
From a maintainability point of view, wouldn't it be more convenient for this loop to rely on a `Particle::reset()` method, instead of copy-pasting the ctor? If we ever add a new struct member or remove an ifdef inside `struct Particle`, and forget about that code duplication in `ghosts.cpp`, the compiler won't generate a warning. This is also exposing too much of the `Particle` implementation details.

Reply:
Actually, this isn't needed at all.

Reply:
In principle no initialization should be needed at all, because the fields should not be accessed before they are updated. In reality multiple methods break if I do this. This is not a good sign.